### PR TITLE
Implementation of statistical analysis for all macro-benchmarks

### PR DESCRIPTION
**Pull request by frouioui**
## Problematic

This Pull Request is the first (big) step towards fixing the very long standing issue in `arewefastyet`: **reliability**. Since the project was created we ran lengthy benchmarks (25-35 mins) and ran them only once per config (SHA + workload + benchmark reason). Leaving us with potentially impaired results and no ability to distinguish the noise from the actual results. It was common for arewefastyet to observe a ±1.5% variation in results for the exact same benchmark, which in the long run is not okay. More information on the problem in https://github.com/vitessio/arewefastyet/issues/515.

## Proposed change

This Pull Request changes the entire benchmarking process: from when we spawn new benchmarks in the queue to when we render results on the UI. Along the way I made some minor refactoring and removed old and/or unused pieces of code such as: notification (was broken and will re-implement it later), some old tests, HTTP argument names, etc.

The _main_ change in this Pull Request is that we are now running the exact same benchmark `MaximumBenchmarkWithSameConfig` times (constant set to `10`) and we now use statistical analysis to observe and compare all of our results. 

### Queue

When the the CRON's handlers spawn new benchmarks they get added to the queue, before being added we multiply the benchmark by `MaximumBenchmarkWithSameConfig` minus the number of same benchmarks already being executed, in the queue and stored in the database as `finished`.

### Execution

After these new benchmarks are added to the queue, they get consumed by the execution component. Now we add two new labels to Ansible: `KeyLastIsSame` and `KeyNextIsSame`, that allow us to optimize how we prepare and clean up the host machine with Ansible. `KeyLastIsSame` is set if the previous benchmark is the same as the current one in which case we don't have to prepare the host, and `KeyNextIsSame` allows us to skip the clean up at the end of the benchmark if the next benchmark is the same. The time spent on sysbench also changed from 600 seconds to 60 seconds, allowing us to re-run a benchmark in less than 2-3 minutes.

Not much else change in the execution.

### Results

The API has changed quite a lot, we used to fetch all the results for a config and doing the median on all the results found. Instead, we now get all the results for a given benchmark and perform simple operations to get some statistics on the results: median, high, low, a 95% confidence interval. If we are just looking to get the results of one benchmark, we return those stats. Otherwise, if we want to compare two benchmarks, we take the two samples and their stats and perform statistical analysis using the Mann Whitney U test. From this, we get a delta in %, a P value and an Alpha value, enabling us to say if we can make statistical inference from the two samples. We are relying on some of the functions used by `benchmath` for this whole process.

### Website

The website has been changed to reflect all the new measurements and changes done to the API. Here is what the new comparison table looks like when comparing `v19.0.1` and a recent commit on Vitess' `main`:

image

## Related Issue

Resolves https://github.com/vitessio/arewefastyet/issues/515

### CODE REVIEW

File: ansible/macrobench.yml [hunk 1]
```diff
@@ -35,7 +38,7 @@
   tags:
     - prometheus
   roles:
-    - { role: cloudalchemy.prometheus, when: stats_remote_db_host is defined }
+    - { role: cloudalchemy.prometheus, when: stats_remote_db_host is defined and arewefastyet_last_exec_is_same is undefined }
```
Comment (frouioui) [inline ansible/macrobench.yml:41]: There is a linter issue here on Ansible, which is fine to ignore. Been a long standing issue.

File: ansible/roles/macrobench/tasks/main.yml [hunk 2]
```diff
@@ -13,6 +13,7 @@
 - name: Install arewefastyet
   become: yes
   become_user: root
+  when: arewefastyet_last_exec_is_same is undefined
```
Comment (notfelineit) [inline ansible/roles/macrobench/tasks/main.yml:16]: This is my first time looking at this, could you explain what this does in layman's terms 😄

File: go/exec/exec.go [hunk 3]
```diff
@@ -156,9 +167,17 @@ const (
 
 // NewExec creates a new *Exec with an autogenerated uuid.UUID as well
 // as a constructed infra.Infra.
```
Comment (notfelineit) [inline go/exec/exec.go:169]: I guess that means this comment can be updated with "autogenerated uuid.UUID OR provided UUID"? (If I understand correctly)

File: go/server/cron.go [hunk 4]
```diff
@@ -57,6 +60,12 @@ var (
 	queue            executionQueue
 )
 
+func (ei executionIdentifier) equalWithoutUUID(id executionIdentifier) bool {
+	ei.UUID = ""
+	id.UUID = ""
```
Comment (notfelineit) [inline go/server/cron.go:65]: Would this permanently set the UUIDs to empty string once `equalWithoutUUID` is called? Don't have the context on whether or not that is intentional!

File: website/components.json [hunk 5]
```diff
@@ -6,7 +6,7 @@
   "tailwind": {
     "config": "tailwind.config.ts",
     "css": "assets/styles/tailwind.css",
-    "baseColor": "neutral",
+    "baseColor": "slate",
```
Comment (notfelineit) [inline website/components.json:9]: 🎨

File: website/src/common/Macrobench.tsx [hunk 6]
```diff
@@ -42,197 +43,217 @@ export default function Macrobench({ data, gitRef, commits }) {
...
+  if (range.infinite == true) {
+    return "∞"
+  }
+  if (range.unknown == true) {
+    return "?"
+  }
+  return "±"+fixed(range.value, 1)+"%"
+}
+
+function Row({ title, oldVal, newVal, delta, insignificant, p, fmt }) {
```
Comment (notfelineit) [inline website/src/common/Macrobench.tsx:218]: This is nit, but since this is a tsx file anyway, could add types to this 😄

File: website/src/pages/ComparePage/ComparePage.tsx [hunk 7]
```diff
@@ -80,65 +72,17 @@ export default function Compare() {
...
+                    commits={{ old: gitRef.old, new: gitRef.new }}
                   />
                 </div>
               );
             })}
           </div>
         </section>
       )}
-
-      {/*{!isMicrobenchLoading && dataMicrobench && (*/}
```
Comment (notfelineit) [inline website/src/pages/ComparePage/ComparePage.tsx:95]: Could we delete this?

**Review by notfelineit — APPROVED** [review]
The math & golang look fine to me! I don't have greater context on arewefastyet though 😄

File: go/server/cron.go (hunk 4 above)
Comment (notfelineit) [reply go/server/cron.go:65]: Ha! You're totally right. I need to brush off more golang.

File: website/src/pages/ComparePage/ComparePage.tsx (hunk 7 above)
Comment (frouioui) [reply website/src/pages/ComparePage/ComparePage.tsx:95]: Yes, this whole comment was not used and not maintained. We disabled **micro** benchmarks a while ago. We should re-think them entirely if we want to re-implement them. This is just some clean up deletion.

File: ansible/roles/macrobench/tasks/main.yml (hunk 2 above)
Comment (frouioui) [reply ansible/roles/macrobench/tasks/main.yml:16]: Sure! On the snippet you highlighted we are checking if the last execution/benchmark _**is not**_ the same (same commit, same source/trigger and same workload). If the condition is true (last benchmark is different than the current one) we would execute the whole `Install arewefastyet` instruction.

The keyword `arewefastyet_last_exec_is_same` is set by the `exec` package when [preparing Ansible](https://github.com/vitessio/arewefastyet/blob/5053153cdaeb44537f44b4f2e32a65e43110eb99/go/exec/exec.go#L372) as an "[extra vars](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_variables.html#defining-variables-at-runtime)". On the snippet below, if the last execution/benchmark **_is_** the same, then we set the "extra vars" `arewefastyet_last_exec_is_same` and then it will not longer be `undefined` when running Ansible and thus the whole `Install arewefastyet` step can be skipped.

```go
        if e.PreviousBenchmarkIsTheSame {
                e.AnsibleConfig.AddExtraVar(ansible.KeyLastIsSame, true)
        }
```

This whole thing is just to optimize how we run Ansible and avoid unnecessary steps, reducing the time needed to setup and cleanup the host machine from 10+ minutes to only 1-2 minutes.

Comment (frouioui) [reply ansible/roles/macrobench/tasks/main.yml:16]: I hope that makes sense, let me know if it is unclear :)

File: website/src/common/Macrobench.tsx (hunk 6 above)
Comment (frouioui) [reply website/src/common/Macrobench.tsx:219]: This is a good point, I am not an expert of TypeScript but I gave it a try on the three `Row` functions we have: [10a00d9](https://github.com/vitessio/arewefastyet/pull/517/commits/10a00d970c2a6eaec3bf158a869a84876c9e664c), [dc4199f](https://github.com/vitessio/arewefastyet/pull/517/commits/dc4199f1b09e58dfa2551a5083778398e5e69ef0) and [058af50](https://github.com/vitessio/arewefastyet/pull/517/commits/058af50dce9554f8b8cb4f7dea40426c41309893)

**Review by vmg — APPROVED** [review]
This looks great to me -- I've only reviewed the Golang code because the React stuff is above my paygrade. I don't think we should block on anything on merging this: the ultimate review will be testing it in practice with real benchmarks. There may be issues that simply cannot be seen by reading the code.

Overall, fantastic work here flo
